README.md
# fyp-test

CHANGELOG.txt
Changes with version 1.0.0 (2016-05-19):
*) Feature: All API methods are now returning a promise instead of using callbacks.
*) Feature: Added new CRUD operations for API endpoints


Gruntfile.js
var srcs = ["lib/relayr.js", "vendor/mqttws31.min.js"];
var ieSrcs = srcs.concat(["vendor/es6-promise.js"]);

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
/*    curl:{
      'vendor/mqttws31.min.js': 'http://git.eclipse.org/c/paho/org.eclipse.paho.mqtt.javascript.git/plain/src/mqttws31.js',
    },*/
    uglify: {
      options: {
        banner: '/*! <%= pkg.name %> <%= grunt.template.today("yyyy-mm-dd") %> */\n'
      },

      build: {
        src: srcs,
        dest: 'build/relayr.min.js'
      },

      buildIe: {
          src: ieSrcs,
          dest: 'build/ie-relayr.min.js'
      }
    },
    concat:{
      dist:{
        src: ["build/relayr.min.js"],
        dest: "build/<%= pkg.name %>.min.js"
      },
      distIe:{
        src: ["build/ie-relayr.min.js"],
        dest: "build/ie-<%= pkg.name %>.min.js"
      }
    },
    jshint: {
      all: ['Gruntfile.js', 'lib/*.js']
    }

  });

  // Load the plugin that provides the "uglify" task.
  grunt.loadNpmTasks('grunt-contrib-concat');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-curl');
  grunt.loadNpmTasks('grunt-contrib-uglify');

  // Default task(s).
  grunt.registerTask('default', ['jshint', 'uglify:build', 'uglify:buildIe', 'concat:dist', 'concat:distIe']);

};


browserSDKFullExample/nodeVirtual.js
//Install mqtt library using: npm install mqtt
var mqtt = require('mqtt');

var client = mqtt.connect({
    servers: [{
        'host': 'mqtt.relayr.io'
    }],
    username: "username",
    password: "password",
    clientId: "clientId",
    protocol: 'mqtts'
});


client.on('connect', function() {

    //subscribe to commands sent from the dashboard or other clients
    client.subscribe("/v1/" + username + "/cmd");

    client.on('message', function(topic, message) {
        console.log(message.toString());
    });

    //simple timer to send a message every 1 second
    var publisher = setInterval(function() {

        // publish a message to a topic
        var data = JSON.stringify({
            meaning: "someMeaning",
            value: "30"
        });

        client.publish("/v1/" + username + "/data", data, function() {
            console.log("Message is published");
        });

    }, 1000);

});

browserSDKFullExample/js-dashboard.js
//connect to cloud
var relayr = RELAYR.init({
    // this comes from the api key page on the dashboard
    //PROJECT ID from Relayr!!
    appId: "fd6eb5e1-681a-414e-83da-5716aa6345e1",
    // this identifies my website as a 'trusted user' basically- it expects me to show up and ask for access to stuff
    redirectUri: "http://jiehui79.github.io/fyp-test/browserSDKFullExample/html-dashboard.html"
});

// these could hold the output from various sensors, based on device id for instance,
// so I can do multiple device calls at once and not get confused
var dev1
var dev2

// in order to do anything other than get straight readings, you have to log in
relayr.login({
    // the login function returns success or error, 
    // the token is generated when you log in to your account in that redirect, 
    // and is passed in the local memory of the browser
    success: function(token) {

        //getUserInfo creates a Promise, so call .then on it
        relayr.user().getUserInfo().then(
            //a fulfilled promise returns an object which has user properties, including email
            function fulfilled(msg) {
                //inject this text into the html
                $(".users").text(msg.email);
            },
            //if the promise resolves as rejected it will log this error
            function rejected(err) {
                console.log("error, the promise was rejected")
            }
        );

        //get all the devices asstd with an account, the loop is just to dispay them
        relayr.devices().getAllDevices().then(
            //a fulfilled promise returns an object msg
            function fulfilled(msg) {
                // loops through the object holding the devices, x gives you an index
                for (x in msg) {
                    // tack the object[index].name on to the list displayed in the html
                    $('<ul>').text(msg[x].name).appendTo('.devices');
                }
            },
            //if the promise resolves as rejected it will log this error
            function rejected(err) {
                console.log("error, the promise was rejected")
            }
        );

        // this gets the data from the devices
        relayr.devices().getDeviceData({
            // this is the same token as above
            token: token,
            // identifies one device from another
            deviceId: "672c5ce6-d170-4278-a68f-8f5e2b32c124",
            //function that grabs the reading from the data from the device
            incomingData: function(data) {
                dev1 = data.readings[0].value;
                //inserts into html
                $(".reading1").text(dev1);
            }
        });

        // same dance, different device
        relayr.devices().getDeviceData({
            token: token,
            deviceId: keys.DEVICE_ID2,
            incomingData: function(data) {
                dev2 = data.readings[0].value
                $(".reading2").text(dev2);
            }
        });

        // displays all of the user's groups
        //get all the devices asstd with an account, the loop is just to dispay them
        relayr.groups().getAllGroups().then(
            //a fulfilled promise returns an object msg
            function fulfilled(msg) {
                // loops through the object holding the devices, x gives you an index
                for (x in msg) {
                    // tack the object[index].name on to the list displayed in the html
                    $('<ul>').text(msg[x].name).appendTo('.groups');
                }
            },
            //if the promise resolves as rejected it will log this error
            function rejected(err) {
                console.log("error, the promise was rejected")
            }
        );

        // displays models available to the user (including public ones)
        relayr.models().getAllModels().then(
            //a fulfilled promise returns an object msg
            function fulfilled(msg) {
                // loops through the object holding the models, x gives you an index
                for (x in msg) {
                    // tack the object[index].name on to the list displayed in the html
                    $('<ul>').text(msg[x].name).appendTo('.models');
                }
            },
            //if the promise resolves as rejected it will log this error
            function rejected(err) {
                console.log("error, the promise was rejected")
            }
        );

        // displays transmitters
        relayr.transmitters().getTransmitters().then(
            //a fulfilled promise returns an object msg
            function fulfilled(msg) {
                // loops through the object holding the transmitters, x gives you an index
                for (x in msg) {
                    // tack the object[index].name on to the list displayed in the html
                    $('<ul>').text(msg[x].name + " : " + msg[x].id).appendTo('.transmitterlist');
                }

                //define what happens when you click the delete button
                $("#delete").click(function() {
                    var transmitter = {
                        transmitterId: String(msg[0].id)
                    };
                    //give the command to actually delete it
                    relayr.transmitters().deleteTransmitter(transmitterId).then(
                        function fulfilled(msg) {
                            location.reload();
                        },
                        function rejected(err) {
                            console.log("error, the promise was rejected")
                        }
                    );
                });

                //define what happens when you click the "update name" button
                $("#updateName").click(function() {
                    //get the ID of the transmitter at the top of the list

                    var transmitter = {
                        transmitterId: String(msg[0].id),
                        name: $('.status-box').val()
                    };
                    //give the command to update the name of the transmitter with the top ID with the text from the input box
                    relayr.transmitters().updateTransmitter(transmitter).then(
                        function fulfilled(msg) {
                            location.reload();
                        },
                        function rejected(err) {
                            console.log("error, the promise was rejected")
                        });
                });

            },

            //if the promise resolves as rejected it will log this error
            function rejected(err) {
                console.log("error, the promise was rejected")
            }
        );

    } //end of success
}); //end of login

browserSDKFullExample/nodeVirtual2.js
//Install mqtt library using: npm install mqtt
var mqtt = require('mqtt');

var client = mqtt.connect({
    servers: [{
        'host': 'mqtt.relayr.io'
    }],
    username: "username",
    password: "password",
    clientId: "clientId",
    protocol: 'mqtts'
});


client.on('connect', function() {

    //subscribe to commands sent from the dashboard or other clients
    client.subscribe("/v1/" + username + "/cmd");

    client.on('message', function(topic, message) {
        console.log(message.toString());
    });

    //simple timer to send a message every 1 second
    var publisher = setInterval(function() {

        // publish a message to a topic
        var val = Math.random() * 10;
        var data = JSON.stringify({
            meaning: "other meaning",
            value: val.toFixed(1)
        });

        client.publish("/v1/" + username + "/data", data, function() {
            console.log("Message is published");
        });

    }, 1000);

});

lib/relayr.js
window.RELAYR = (function(win, doc) {
    var relayr = {
        init: function(credentials) {
            return new Relayr(credentials);
        }
    };

    function Relayr(credentials) {


        var config = {
            url: 'https://api.relayr.io/',
            mqtt: 'mqtt.relayr.io',
            mockMode: false
        };


        var self = this;

        self.credentials = credentials;
        self.config = config;
        self.checkAccount = function() {
            if (!self.account) {
                throw new Error('You must be logged in to access this method.');
            }
        };

        function login(callback) {

            if (!(credentials && util.nonEmptyString(credentials.appId))) {
                throw new Error('Provide credentials: appId and redirectUri');
            }

            if (!(credentials && util.nonEmptyString(credentials.redirectUri))) {
                throw new Error('Provide credentials: redirectUri');
            }
            if (!(callback && callback.success && typeof callback.success == 'function')) {
                throw new Error('Provide the method success within your parameters');
            }
            var localToken = localStorage.getItem('relayrToken');
            if (localToken) {
                triggerCallback();
            } else {

                var authURL = {
                    client_id: self.credentials.appId,
                    redirect_uri: self.credentials.redirectUri,
                    scope: 'access-own-user-info+configure-devices'
                };

                this.login.redirect(config.url + 'oauth2/auth?client_id=' + authURL.client_id + '&redirect_uri=' + authURL.redirect_uri + '&response_type=token&scope=' + authURL.scope);
            }

            function triggerCallback() {
                self.user().getUserInfo(localToken).then(
                    function() {
                        callback.success(localToken);
                    }, function(error) {
                        if (!(callback && callback.error && typeof callback.error == 'function')) {
                            localStorage.removeItem('relayrToken');
                            self.login(callback);
                        } else {
                            callback.error(error);
                        }
                    });
            }
        }

        function user() {
            this.getToken = function() {
                //Auto check for token scrapping
                function parse(input) {
                    var parts = input.split('#');
                    if (parts.length < 2) {
                        return undefined;
                    }
                    var queryParams = parts[1].split('&');
                    var obj = queryParams.reduce(function(accumulator, pair) {
                        var tuple = pair.split('=');
                        accumulator[tuple[0]] = tuple[1];
                        return accumulator;
                    }, {});
                    return obj.access_token;
                }
                var parsed = parse(util.getHash());
                var loc = window.location.href,
                    index = loc.indexOf('#access_token');

                if (index > 0) {
                    window.location = loc.substring(0, index);
                }
                if (parsed) {
                    this.setToken(parsed);
                }
            };

            this.setToken = function(token) {
                localStorage.setItem('relayrToken', token);
            };

            this.hasToken = function() {
                return !!localStorage.getItem('relayrToken');
            };

            this.getUserInfo = function(token) {
                return new Promise(function(resolve, reject) {
                    if (self.account) {
                        resolve(self.account);
                    } else {
                        util.ajax({
                            url: self.config.url + 'oauth2/user-info',
                            type: 'GET',
                            token: 'Bearer ' + token,
                            isObject: true
                        }, function(account) {
                            account.token = token;
                            self.account = account;
                            resolve(self.account);
                        }, function(error) {
                            reject(error);
                        });
                    }
                });
            };

            this.logout = function() {
                localStorage.removeItem('relayrToken');
            };
        }



        function devices() {
            this.getDeviceData = function(device) {
                return new Promise(function(resolve, reject) {

                    if (!device.token) {
                        if (self.account.token) {
                            device.token = self.account.token;
                        }
                    }

                    device.token = device.token.replace('Bearer ', '');
                    util.ajax({
                        url: self.config.url + 'channels',
                        type: 'POST',
                        token: 'Bearer ' + device.token,
                        isObject: true,
                        body: {
                            'transport': 'mqtt',
                            'deviceId': device.deviceId
                        }
                    }, function(credentials) {
                        credentials.cb = device.incomingData;
                        util.mqttClient(credentials);
                    }, function(error) {
                        reject(error);
                        localStorage.removeItem('relayrToken');
                        self.login();
                    });
                });
            };

            this.getDevice = function(device) {
                if (!(device && device.deviceId)) {
                    throw new Error('Provide the deviceId within your parameters');
                }

                return new Promise(function(resolve, reject) {
                    util.ajax({
                        url: self.config.url + 'devices/' + device.deviceId,
                        type: 'GET',
                        token: 'Bearer ' + self.account.token,
                        isObject: true
                    }, function(device) {
                        resolve(device);
                    }, function(error) {
                        reject(error);
                    });
                });
            };

            this.getDeviceState = function(device) {
                if (!(device && device.deviceId)) {
                    throw new Error('Provide the deviceId within your parameters');
                }

                return new Promise(function(resolve, reject) {
                    util.ajax({
                        url: self.config.url + 'devices/' + device.deviceId + '/state',
                        type: 'GET',
                        token: 'Bearer ' + self.account.token,
                        isObject: true
                    }, function(device) {
                        resolve(device);
                    }, function(error) {
                        reject(error);
                    });
                });
            };

            this.getAllDevices = function() {
                return new Promise(function(resolve, reject) {
                    util.ajax({
                        url: self.config.url + 'users/' + self.account.id + '/devices',
                        type: 'GET',
                        token: 'Bearer ' + self.account.token,
                        isObject: true
                    }, function(devices) {
                        resolve(devices);
                    }, function(error) {
                        reject(error);
                    });
                });
            };

            this.sendCommand = function(device) {
                if (!(device && device.deviceId && device.command)) {
                    throw new Error('Provide the deviceId and a command within your parameters');
                }
                if (!device.token) {
                    if (self.account.token) {
                        device.token = self.account.token;
                    }
                }

                device.token = device.token.replace('Bearer ', '');


                return new Promise(function(resolve, reject) {
                    util.ajax({
                        url: self.config.url + 'devices/' + device.deviceId + '/commands',
                        type: 'POST',
                        token: 'Bearer ' + device.token,
                        body: device.command
                    }, function(credentials) {
                        resolve(credentials);
                    }, function(error) {
                        reject(error);
                        localStorage.removeItem('relayrToken');
                        self.login();
                    });
                });
            };
        }

        function groups() {
            this.getAllGroups = function() {
                return new Promise(function(resolve, reject) {
                    util.ajax({
                        url: self.config.url + 'users/' + self.account.id + '/groups',
                        type: 'GET',
                        token: 'Bearer ' + self.account.token,
                        isObject: true
                    }, function(group) {
                        resolve(group);
                    }, function(error) {
                        reject(error);
                    });
                });
            };
        }

        function models() {
            this.getAllModels = function() {
                return new Promise(function(resolve, reject) {
                    util.ajax({
                        url: self.config.url + 'device-models?limit=100000',
                        type: 'GET',
                        token: 'Bearer ' + self.account.token,
                        isObject: true
                    }, function(group) {
                        resolve(group);
                    }, function(error) {
                        reject(error);
                    });
                });
            };
        }

        function transmitters() {
            self.checkAccount();

            this.getTransmitters = function() {
                return new Promise(function(resolve, reject) {
                    util.ajax({
                        url: self.config.url + 'users/' + self.account.id + '/transmitters',
                        type: 'GET',
                        token: 'Bearer ' + self.account.token,
                        isObject: true
                    }, function(transmitters) {
                        resolve(transmitters);
                    }, reject);
                });
            };

            this.deleteTransmitter = function(transmitter) {
                if (!(transmitter && transmitter.transmitterId)) {
                    throw new Error('Provide the transmitterId within your parameters');
                }
                return new Promise(function(resolve, reject) {
                    util.ajax({
                        url: self.config.url + "transmitters/" + transmitter.transmitterId,
                        type: "DELETE",
                        token: "Bearer " + self.account.token
                    }, function(transmitters) {
                        resolve(transmitters);
                    }, reject);
                });
            };

            this.updateTransmitter = function(transmitter) {
                if (!(transmitter && transmitter.transmitterId)) {
                    throw new Error('Provide the transmitterId within your parameters');
                }
                return new Promise(function(resolve, reject) {
                    util.ajax({
                        url: self.config.url + "transmitters/" + transmitter.transmitterId,
                        type: "PATCH",
                        token: "Bearer " + self.account.token,
                        body: transmitter,
                        isObject: true
                    }, function(transmitters) {
                        resolve(transmitters);
                    }, reject);
                });
            };
        }


        this.login = login;
        this.user = function() {
            return new user();
        };

        this.device = function(deviceObj) {
            return new device(deviceObj);
        };

        this.devices = function() {
            return new devices();
        };

        this.groups = function() {
            return new groups();
        };

        this.models = function() {
            return new models();
        };

        this.transmitters = function() {
            return new transmitters();
        };

        this.login.redirect = function(uri) {
            doc.location = uri;
        };

        this.util = util;

        (function() {
            this.user().getToken();
        }.bind(this))();
    }

    var util = {
        nonEmptyString: function(string) {
            return string && typeof string == 'string' && string.length > 0;
        },

        isAFunction: function(fn) {
            return fn && typeof fn == 'function';
        },

        getHash: function() {
            return document.location.hash;
        },

        ajax: function(options, callback, error) {
            var xhrObject = new XMLHttpRequest();
            xhrObject.onreadystatechange = function() {
                if (xhrObject.readyState === 4) {
                    if (xhrObject.status > 199 && xhrObject.status < 299) {

                        if (options.isObject) {
                            callback(JSON.parse(xhrObject.responseText));
                        } else {
                            callback(xhrObject.responseText);
                        }
                    }
                    if (xhrObject.status > 399 && xhrObject.status < 600) {
                        error(xhrObject);
                    }
                }
            };
            xhrObject.open(
                options.type,
                options.url,
                true
            );

            xhrObject.setRequestHeader('Authorization', options.token);
            xhrObject.setRequestHeader('Content-Type', 'application/json');
            if (options.body) {
                xhrObject.send(JSON.stringify(options.body));
            } else {
                xhrObject.send();
            }
        },

        mqttClient: function(credentials) {
            credentials.mqtt = new mqtt();
            credentials.mqtt.subscribe(credentials);
            credentials.mqtt.connect();
            return credentials;
        }
    };

    function mqtt(base) {
        var MQTT_TIMEOUT = 10000;
        var self = this;
        self.subscriptions = [];
        self.credentials = null;

        self.retries = 0;
        self.connect = function(param) {
            param = self.subscriptions[0];
            self.credentials = param;
            self.client = new Paho.MQTT.Client('mqtt.relayr.io', 443, 'JSDK_' + Math.floor((Math.random() * 1000)));
            self.client.onConnectionLost = onConnectionLost;
            self.client.onMessageArrived = onMessageArrived;
            self.client.isConnected = false;
            self.retries++;
            console.log('Connecting...');
            if (self.retries > 100) {
                throw new Error('Too many MQTT retries occured, please try later.');
            }

            var options = {
                timeout: 30,
                keepAliveInterval: 10,
                cleanSession: true,
                useSSL: true,
                userName: param.credentials.user,
                password: param.credentials.password,
                onSuccess: function(data) {
                    self.client.isConnected = true;
                    self.retries = 0;
                    self.subscriptions.forEach(function(subscription) {
                        self.client.subscribe(subscription.credentials.topic, 0);

                    });
                },
                onFailure: function(data) {
                    if (!self.forceDisconnect) {
                        self.connect(self.credentials);
                    }
                }
            };

            self.client.connect(options);
        };


        function onMessageArrived(data) {
            try {
                var deviceId = data._getDestinationName();
                deviceId = deviceId.split('/v1/')[1].split('/')[0];
                var dataTopic = data._getDestinationName().split('/v1/')[1];
                var incomingData = (data._getPayloadString());
                incomingData = JSON.parse(data._getPayloadString());

                for (var i = self.subscriptions.length - 1; i >= 0; i--) {
                    if (self.subscriptions[i].channelId == dataTopic) {
                        deviceId = self.subscriptions[i].deviceId;

                        if ((self.subscriptions[i].cb && self.subscriptions[i].cb && typeof self.subscriptions[i].cb == 'function')) {
                            self.subscriptions[i].cb(incomingData);
                        }
                        if ((self.subscriptions[i].incomingData && self.subscriptions[i].incomingData && typeof self.subscriptions[i].incomingData == 'function')) {
                            self.subscriptions[i].incomingData(incomingData);
                        }
                    }
                }

            } catch (err) {
                console.log('Incoming data function error:', err);
            }
        }

        function onConnectionLost(data) {
            console.log('Lost connection');
            setTimeout(function() {
                self.client.isConnected = false;
                self.connect(self.lastParam);

                if (!self.forceDisconnect) {
                    self.connect(self.lastParam);
                }
            }, MQTT_TIMEOUT);
        }

        this.disconnect = function() {
            self.forceDisconnect = true;
            if (self.client) {
                try {
                    self.client.disconnect();
                } catch (err) {}
            }
        };

        this.subscribe = function(channel) {
            self.subscriptions.push(channel);
        };

        this.cleanUp = function() {
            self.disconnect();
        };

        return this;
    }

    return relayr;
}(window, window.document));
